File: src/api/tags.ts
import api from './client';
import { Tag } from '../types';

export interface TagPayload {
  name: string;
  color: string;
}

export const tagsApi = {
  list: () => api.get<Tag[]>('/tags').then(r => r.data),
  get: (id: string) => api.get<Tag>(`/tags/${id}`).then(r => r.data),
  create: (data: TagPayload) => api.post<Tag>('/tags', data).then(r => r.data),
  update: (id: string, data: Partial<TagPayload>) =>
    api.put<Tag>(`/tags/${id}`, data).then(r => r.data),
  delete: (id: string) => api.delete(`/tags/${id}`).then(r => r.data),
};

File: backend/routers/tags.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
import uuid
from database import get_db
from models import Tag
from schemas import TagCreate, TagUpdate, TagRead

router = APIRouter()


@router.get("", response_model=List[TagRead])
def list_tags(db: Session = Depends(get_db)):
    return db.query(Tag).order_by(Tag.name).all()


@router.post("", response_model=TagRead)
def create_tag(data: TagCreate, db: Session = Depends(get_db)):
    tag = Tag(
        id=str(uuid.uuid4()),
        name=data.name,
        color=data.color,
    )
    db.add(tag)
    db.commit()
    db.refresh(tag)
    return tag


@router.get("/{tag_id}", response_model=TagRead)
def get_tag(tag_id: str, db: Session = Depends(get_db)):
    tag = db.query(Tag).filter(Tag.id == tag_id).first()
    if not tag:
        raise HTTPException(status_code=404, detail="Tag not found")
    return tag


@router.put("/{tag_id}", response_model=TagRead)
def update_tag(tag_id: str, data: TagUpdate, db: Session = Depends(get_db)):
    tag = db.query(Tag).filter(Tag.id == tag_id).first()
    if not tag:
        raise HTTPException(status_code=404, detail="Tag not found")
    for field, value in data.model_dump(exclude_none=True).items():
        setattr(tag, field, value)
    db.commit()
    db.refresh(tag)
    return tag


@router.delete("/{tag_id}")
def delete_tag(tag_id: str, db: Session = Depends(get_db)):
    tag = db.query(Tag).filter(Tag.id == tag_id).first()
    if not tag:
        raise HTTPException(status_code=404, detail="Tag not found")
    db.delete(tag)
    db.commit()
    return {"ok": True}
